lightning/src/ln/channelmanager.rs

```diff
@@ -2809,6 +2852,7 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 						//TODO: Resend the funding_locked if needed once we get the monitor running again
 					}
 				}
+				if let Err(e) = self.fail_holding_cell_htlcs(htlcs_to_fail, msg.channel_id) { return Err(e) };
```

Comment:
This looks like a deadlock - you're calling fail_holding_cell_htlcs with channel_state locked, but then fail_holding_cell_htlcs will try to lock channel_state if it has work to do. You could address this in one of two ways - either pass a reference to the channel_state lock to fail_holding_cell_htlcs, or return the htlcs_to_fail outside of the match and fail after dropping the lock. May be best to do the second one if you can avoid channel_reestablish returning early and us forgetting HTLCs because it'll make future channel_state refactors easier.

Reply:
Thanks, do you think it's a good idea to add a test for the "on channel_reestablish, fail holding cell htlcs" case? I'm not sure how that's gonna work exactly but looking into it now..